src/ray/raylet/scheduling/policy/node_label_scheduling_policy.cc

```diff
@@ -43,26 +43,40 @@ scheduling::NodeID NodeLabelSchedulingPolicy::Schedule(
   }
 
   // 3. Filter by soft expressions.
+  absl::flat_hash_map<scheduling::NodeID, const Node *> soft_match_nodes;
   const auto &soft_expressions = node_label_scheduling_strategy.soft();
   if (soft_expressions.expressions().size() > 0) {
-    auto match_nodes =
+    soft_match_nodes =
         FilterNodesByLabelMatchExpressions(candidate_nodes, soft_expressions);
-    if (!match_nodes.empty()) {
-      candidate_nodes = match_nodes;
-    }
   }
 
-  return SelectBestNode(candidate_nodes, resource_request);
+  return SelectBestNode(candidate_nodes, soft_match_nodes, resource_request);
 }
 
 scheduling::NodeID NodeLabelSchedulingPolicy::SelectBestNode(
     const absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const absl::flat_hash_map<scheduling::NodeID, const Node *> &soft_match_nodes,
```

Comment:
maybe rename to something like `hard_match_nodes` and `hard_and_soft_match_nodes`?

Reply:
Fixed